### PR TITLE
Loading: convert transposed and RoPE-permuted weights before taking the DTensor shard

### PR DESCRIPTION
**Pull request by qgallouedec**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48519) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48519)

When a parameter is loaded as a DTensor (`tp_plan`, or FSDP2 at load time), the loader slices the checkpoint tensor to this rank's shard first and runs the conversion ops on the slice. `Transpose` and `PermuteForRope` move elements across dims, so when the sharded dim is one they touch, the shard of the source is not the shard of the converted tensor:

- `Transpose`: a router stored as `(hidden, experts)`, sharded on dim 0 over 2 ranks, comes out as `(experts, hidden/2)` against a `(experts/2, hidden)` shard. Loud: mismatched-keys error.
- `PermuteForRope`: takes the head size from `tensor.shape[0]`, i.e. from the shard. Silent: q/k permuted with the wrong head size, no error.

Repro (Kimi K2.5 vision q/k go through `PermuteForRope`; needs FSDP2 at load, #48205):

First:
```python
import torch
from transformers import Kimi_K25ForConditionalGeneration
from tests.models.kimi_k25.test_modeling_kimi_k25 import Kimi_K25VisionText2TextModelTester

model = Kimi_K25ForConditionalGeneration(Kimi_K25VisionText2TextModelTester(None).get_config())
model.save_pretrained(path, save_original_format=True)  # q/k written un-permuted, permuted back at load
torch.save(model.state_dict(), "/tmp/kimi_tiny/plain.pt")
```

```python
# torchrun --nproc_per_node 2 repro.py
import torch
from transformers import Kimi_K25ForConditionalGeneration
from transformers.distributed import DistributedConfig

model = Kimi_K25ForConditionalGeneration.from_pretrained(path, distributed_config=DistributedConfig(fsdp_size=2))
plain = torch.load("/tmp/kimi_tiny/plain.pt")
for name, param in model.named_parameters():
    if "vision" in name and ("q_proj.weight" in name or "k_proj.weight" in name):
        print(name, (param.full_tensor().cpu() - plain[name]).abs().max().item())
```

Before (no error raised):
```
model.vision_tower.layers.0.attn.q_proj.weight 0.0897
model.vision_tower.layers.0.attn.k_proj.weight 0.1171
model.vision_tower.layers.1.attn.q_proj.weight 0.0956
model.vision_tower.layers.1.attn.k_proj.weight 0.0965
```

After:
```
model.vision_tower.layers.0.attn.q_proj.weight 0.0
model.vision_tower.layers.0.attn.k_proj.weight 0.0
model.vision_tower.layers.1.attn.q_proj.weight 0.0
model.vision_tower.layers.1.attn.k_proj.weight 0.0
```

Fix:
- `shards_after_conversion` says whether a converter has one of these ops on a sharded dim;
- for those the full tensor is converted, then sharded with the same `DtensorShardOperation`. 

Everything else keeps shard-on-read (fused experts, and transposes on unsharded dims such as Qwen3-VL-MoE's)

### CODE REVIEW

**Comment by vasqu** [comment]
cc @3outeille could you take a first look here?

File: src/transformers/core_model_loading.py [hunk 1]
```diff
@@ -1269,6 +1269,20 @@ def _job():
...
+    taken: a transpose or a RoPE permutation on a sharded dim makes the shard of the source the wrong slice."""
+    if not isinstance(mapping, WeightConverter):
+        return False
+    shard_dims = {placement.dim % ndim for placement in placements if placement.is_shard()}
+    for op in mapping.operations:
+        if isinstance(op, Transpose) and {op.dim0 % ndim, op.dim1 % ndim} & shard_dims:
+            return True
+        if isinstance(op, PermuteForRope) and 0 in shard_dims:
+            return True
+    return False
```
Comment (vasqu) [inline src/transformers/core_model_loading.py:1283]: Wondering whether we might move this as general function within converter/renaming and override where we need it

Would make it easier to scale tbh

**Review by vasqu — COMMENTED** [review]
Just one initial design question but letting Ferdinand comment first on the logic

File: src/transformers/core_model_loading.py (hunk 1 above)
Comment (qgallouedec) [reply src/transformers/core_model_loading.py:1283]: Yes I wasn't sure. I went for simplicity instead of generality as a first step, but happy to generalize

Comment (vasqu) [reply src/transformers/core_model_loading.py:1283]: I just fear the function exploding at some point :cry: and it makes custom extensions simpler

File: src/transformers/core_model_loading.py [hunk 2]
```diff
@@ -1718,6 +1732,10 @@ def convert_and_load_state_dict_in_model(
             sharding_op = None
             if is_dtensor(empty_param):
                 sharding_op = DtensorShardOperation(empty_param)
+            # A transpose or a RoPE permutation moves elements across the sharded dim, so this rank's slice of the
+            # source is not the slice of the converted tensor: those are converted in full and sharded afterwards.
+            if sharding_op is not None and shards_after_conversion(mapping, empty_param.placements, empty_param.ndim):
+                sharding_op = None
```
Comment (VI-Arthur) [inline src/transformers/core_model_loading.py]: you can compute that earlier to avoid initializing a dtensor sharding op.

**Review by VI-Arthur — COMMENTED** [review]
I'll have a deeper look / need to bench a bit but here are the red flags for me:
🔴 Each GPU gets the full tensor, and full conversion: this does not scale / does not work when you have restricted memory
🔴 Each GPU will be doing full computation without any comp/coms overlap.

Even as a temporary solution I don't like it! 
I'll spend some time on it

**Comment by qgallouedec** [comment]
Two things first, because I think the scope is smaller than it looks.

It's not every tensor: `shards_after_conversion` only returns `True` when a `Transpose` touches a sharded dim, or a `PermuteForRope` sits on a sharded dim 0. Everything else keeps shard-on-read.

And the current behaviour isn't "fast", it's wrong. `PermuteForRope` reads the head size from `tensor.shape[0]`, so on a shard it permutes with the wrong head size and no error is raised.

So the comparison isn't slow-but-correct vs fast-but-correct.

Measured, `torchrun --nproc_per_node 2`, same tree both arms (the off arm forces `shards_after_conversion` to `False`, i.e. main today):

| case | params taking the path | full tensor materialized |
|---|---|---|
| Llama-3.2-3B, fsdp=2 | 0 of 508 | none |
| router `gate.weight`, `Transpose(0, 1)` | 1 per layer | 1-2 MiB |
| vision q/k, `PermuteForRope` | 2 per layer | attention-sized |
| `qwen3_vl_moe` experts, `Transpose(1, 2)`, TP on the inner dims | 2 per layer | 0.75 GiB (30B-A3B), 3.0 GiB (235B-A22B) |

Llama-3.2-3B: load 81.5 s, host peak 7.6 GiB, device peak 2.99 GiB, unchanged with the fix off since nothing takes the path.

The expert row is your point and I agree with it, 3 GiB per layer per rank is not something I want either.

On your all-to-all idea: agreed it's better, and iiuc, it could be even better than you think: for a pure transpose the output shard is a strided read of the source, so you can keep shard-on-read entirely, no gather and no full materialization. Same for the RoPE permutation.

The question is whether it blocks this one. If the numbers above are acceptable I'd land this as the correctness fix and do the strided path as a follow-up.

**Comment by VI-Arthur** [comment]
Related to https://github.com/huggingface/transformers/pull/48373 as well which should fix transpose. I agree that given the rarity we should aim for simpler solution: comput the transition map, compute th operations with the transition

**Comment by qgallouedec** [comment]
#48373 nice!, dropped the transpose case.
You're right it doesn't need converting in full

What's left is `PermuteForRope`, which isn't the same shape of problem: it interleaves rows and takes the head size from `tensor.shape[0]`, which on a shard is the local row count, so it permutes wrong and raises nothing!

So this is one function and q/k projections now, which are small. Happy for it to be superseded if you'd rather fold RoPE into the same mechanism.

**Comment by VI-Arthur** [comment]
I'm having a look don't worry!

File: src/transformers/core_model_loading.py (hunk 2 above)
Comment (qgallouedec) [reply src/transformers/core_model_loading.py]: Done in cb61b1aab8, it is only built when it is going to be used:

```python
sharding_op = None
if is_dtensor(empty_param) and not mapping.shards_after_conversion(empty_param.placements, empty_param.ndim):
    sharding_op = DtensorShardOperation(empty_param)
```

File: src/transformers/core_model_loading.py (hunk 1 above)
Comment (qgallouedec) [reply src/transformers/core_model_loading.py]: Went with a method in 071db77579, tell me if that's not what you had in mind.

`shards_after_conversion` is now a method: the default on `ConversionOps` returns `False`, `PermuteForRope` overrides it, `WeightConverter` is `any(op.shards_after_conversion(...) for op in self.operations)` and `WeightTransform` returns `False`. So a new op that breaks shard-on-read declares it itself instead of being added to an isinstance chain in a central function!
